### PR TITLE
feat(swipecontrol): Add the SwipeControl

### PR DESCRIPTION
**Pull request by dr1rrb**
## Feature
Add the `SwipeControl`

## What is the current behavior?
`SwipeControl` is not supported

## What is the new behavior?
🙃

## PR Checklist
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [in progress] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [ ] ~~Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).~~
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/Controls/Icon/FontIconSource.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
+using Windows.UI.Text;
+using Windows.UI.Xaml.Media;
+
+namespace Windows.UI.Xaml.Controls
+{
+	public partial class FontIconSource : global::Windows.UI.Xaml.Controls.IconSource
```
Comment (MartinZikmund) [inline src/Uno.UI/UI/Xaml/Controls/Icon/FontIconSource.cs:6]: We have also the MUX version of these here - `Uno.UI\Microsoft\UI\Xaml\Controls\IconSource`, maybe the code could be copied from there for the other icon types too

Comment (dr1rrb) [reply src/Uno.UI/UI/Xaml/Controls/Icon/FontIconSource.cs:6]: Actually the code is automatically "copied" to MUX (that's why my build failed :) )

File: src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeItem.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+// https://github.com/microsoft/microsoft-ui-xaml/blob/307bd99682cccaa128483036b764c0b7c862d666/dev/SwipeControl/SwipeItem.cpp
+
+using System;
+using System.Windows.Input;
+using Windows.UI.Xaml.Input;
+using Windows.UI.Xaml.Media;
+
+namespace Windows.UI.Xaml.Controls
+{
+	public partial class SwipeItem : FrameworkElement
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeItem.cs:17]: Is it necessary to inherit from `FrameworkElement`? Just noticed that it should really just be a `DependencyObject` (actually when I was wondering about the `new Background` stuff).

If it's necessary, add an 'Uno workaround' comment and an explanation.

File: src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeItems.h.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+		//winrt.IVectorView<winrt.SwipeItem> GetView();
+		//bool IndexOf(winrt.SwipeItem & value, uint32_t& index);
+		//void SetAt(uint index, winrt.SwipeItem & value);
+		//void InsertAt(uint index, winrt.SwipeItem & value);
+		//void RemoveAt(uint index);
+		//void Append(winrt.SwipeItem & value);
+		//void RemoveAtEnd();
+		//void Clear();
+
+		// TODO:
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeItems.h.cs:31]: I don't think any of these weird pseudo-Linq methods are actually part of the .NET projection, you can just remove them.

```csharp
#region Assembly Windows.Foundation.UniversalApiContract, Version=8.0.0.0, Culture=neutral, PublicKeyToken=null, ContentType=WindowsRuntime
// C:\Program Files (x86)\Windows Kits\10\References\10.0.18362.0\Windows.Foundation.UniversalApiContract\8.0.0.0\Windows.Foundation.UniversalApiContract.winmd
#endregion

using System.Collections.Generic;
using Windows.Foundation;
using Windows.Foundation.Metadata;

namespace Windows.UI.Xaml.Controls
{
	//
	// Summary:
	//     Represents a collection of SwipeItem objects.
	[Composable(typeof(ISwipeItemsFactory), CompositionType.Public, 327680, "Windows.Foundation.UniversalApiContract")]
	[ContractVersion(typeof(UniversalApiContract), 327680)]
	[MarshalingBehavior(MarshalingType.Agile)]
	[Static(typeof(ISwipeItemsStatics), 327680, "Windows.Foundation.UniversalApiContract")]
	[Threading(ThreadingModel.Both)]
	[WebHostHidden]
	public class SwipeItems : DependencyObject, ISwipeItems, IList<SwipeItem>, IEnumerable<SwipeItem>
	{
		//
		// Summary:
		//     Initializes a new instance of the SwipeItems class.
		public SwipeItems();

		//
		// Summary:
		//     Gets or sets a value that indicates the effect of a swipe interaction.
		//
		// Returns:
		//     A value of the enumeration that indicates the effect of a swipe interaction.
		//     The default is **Reveal**.
		public SwipeMode Mode { get; set; }
		//
		// Summary:
		//     Identifies the Mode dependency property.
		//
		// Returns:
		//     The identifier for the Mode dependency property.
		public static DependencyProperty ModeProperty { get; }
	}
```

**Review by davidjohnoliver — APPROVED** [review]

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/IconSource/IconSource.cs [hunk 4]
```diff
@@ -13,5 +14,10 @@ public Brush Foreground
 
 		public static DependencyProperty ForegroundProperty { get; } =
 			DependencyProperty.Register(nameof(Foreground), typeof(Brush), typeof(IconSource), new PropertyMetadata(null));
+
+#nullable enable
+
+		internal virtual IconElement? CreateIconElement()
+			=> default;
```
Comment (jeromelaban) [inline src/Uno.UI/Microsoft/UI/Xaml/Controls/IconSource/IconSource.cs:21]:
```suggestion
#nullable enable

		internal virtual IconElement? CreateIconElement()
			=> default;
			
#nullable restore
```

File: src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeControl.Uno.cs [hunk 5]
```diff
@@ -0,0 +1,576 @@
...
+					DeltaT = left.DeltaT + right.DeltaT
+				};
+			}
+		}
+	}
+
+	/// <summary>
+	/// Extension methods for classes in the Windows.UI.Xaml.Media.Animation namespace.
+	/// </summary>
+	internal static class StoryboardExtensions
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeControl.Uno.cs:556]: Should be moved if not part of the original source

File: src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeControl.cs [hunk 6]
```diff
@@ -0,0 +1,1893 @@
...
+					if (globalTestHooks is {})
+					{
+						globalTestHooks.NotifyLastInteractedWithSwipeControlChanged();
+					}
+				}
+			}
+		}
+
+		#region ISwipeControl
+		public async void Close()
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeControl.cs:85]: Async void is likely to cause race conditions when calling it multiple times. We should wrap it and keep track so we at least queue operations properly.

File: src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeControl.cs [hunk 7]
```diff
@@ -0,0 +1,1893 @@
...
+			if (!double.IsInfinity(availableSize.Height))
+			{
+				contentDesiredSize.Height = availableSize.Height;
+			}
+
+			return contentDesiredSize;
+		}
+		#endregion
+
+		#region IInteractionTrackerOwner
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeControl.cs:226]: Use #if false instead to make comparison with the original code easier.
```suggestion
#if false
		#region IInteractionTrackerOwner
```

File: src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeItem.cs [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+using System.Windows.Input;
+using Windows.Devices.Input;
+using Windows.UI.Xaml.Input;
+using Windows.UI.Xaml.Media;
+
+namespace Windows.UI.Xaml.Controls
+{
+	public partial class SwipeItem : FrameworkElement
+	{
+		//static double s_swipeItemWidth = 68.0;
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeItem.cs:20]: Was it unused in the original source ?

File: src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeItems.h.cs (hunk 3 above)
Comment (dr1rrb) [reply src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeItems.h.cs:31]: Actually they are part of the public API ... https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.swipeitems.first?view=winrt-19041#Windows_UI_Xaml_Controls_SwipeItems_First

Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeItems.h.cs:31]: They sure are, if you're using C++. That's why I specified 'the .NET projection' ;)

I was going to copy-paste the actual info from the binary but I see I already did, so... see above...

File: src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeItem.cs (hunk 8 above)
Comment (dr1rrb) [reply src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeItem.cs:20]: I search in the whole original source repo, apparently it wasn't used.

File: src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeItem.cs (hunk 2 above)
Comment (dr1rrb) [reply src/Uno.UI/UI/Xaml/Controls/SwipeControl/SwipeItem.cs]: Good point. Not even sure why it was a `FrameworkElement` ... most probably for bindings, but it was not a good reason so I've converted it to a DO.

**Comment by pdecostervgls** [comment]
I guess this closes #1402 ?

**Pull request merged**